src/runtime/mod.rs

```diff
@@ -452,6 +440,59 @@ impl Module {
         headers
     }
 
+    /// Internal utility function for parsing a host header.
+    ///
+    /// This attempts to use three sources to construct a definitive host/port pair, ordering
+    /// by precedent.
+    ///
+    /// - The content of the host header is considered most authoritative.
+    /// - Next most authoritative is self.host, which is set at the CLI or in the config
+    /// - As a last resort, we use the host/port that Hyper gives us.
+    /// - If none of these provide sufficient data, which is definitely a possiblity,
+    ///   we go with `localhost` as host and `80` as port. This, of course, is problematic,
+    ///   but should only manifest if both the server and the client are behaving badly.
+    fn parse_host_header_uri(&self, headers: &HeaderMap, uri: &hyper::Uri) -> (String, String) {
+        let host_header = headers.get(HOST).and_then(|v| match v.to_str() {
+            Err(_) => None,
+            Ok(s) => Some(s.to_owned()),
+        });
+
+        let mut host = uri
+            .host()
+            .map(|h| h.to_string())
+            .unwrap_or_else(|| "localhost".to_owned());
+        let mut port = uri.port_u16().unwrap_or(80).to_string();
+
+        let mut parse_host = |hdr: String| {
```

Comment:
Any reason not to just split this out into its own function rather than capturing and mutating `host`?

Reply:
Please note, this is a minor nit, so no need to address or re-review

Reply:
Really, the closure was just a way to keep the mutation compartmentalized and DRY. Because we can get port from one source and host from another, it makes sense to mutate host and port. I tried a few variations, and this one seemed to be the least lines of code/allocations.